### PR TITLE
Update PHPJS

### PR DESCRIPTION
**Pull request by badlee**
**Changes**
- Set version to 1.1.2
- Support PHP closures
- Pass parameters to PHP functions
- Support PHP objects
- JS Modules Loading via PHP function
- Secure PHP<->JS bridge

### CODE REVIEW

File: phpjs.c [hunk 1]
```diff
@@ -1,317 +1,368 @@
...
+    SET_PROP(tc_ex, phpjs_JSException_ptr, "message", message);
+    SET_PROP(tc_ex, phpjs_JSException_ptr, "js_stack", js_stack);
+
+    zend_throw_exception_object(tc_ex TSRMLS_CC);
+
+    efree(js_stack);
+    efree(message);
+}
+
+void zval_to_duk(duk_context * ctx, char * name, zval * value)
```
Comment (crodas) [inline phpjs.c:55]: You need to add `TSRMLS_DC`

File: phpjs.c [hunk 2]
```diff
@@ -1,317 +1,368 @@
...
+        duk_push_string(ctx, Z_EXCEPTION_PROP("message"));
+        zend_clear_exception(TSRMLS_C);
+    }
+    if (catch)
+        duk_throw(ctx);
+    //php_printf("End Exec ZEND_FNC WITH %u\n",Z_TYPE_P(retval));
+    
+    if(!catch && returnVal){
+        returnVal = Z_TYPE_P(retval) == 0 ? 0 : 1;
+        zval_to_duk(ctx, NULL, retval);
```
Comment (crodas) [inline phpjs.c:315]: TSRMLS_CC at the end

File: phpjs.c [hunk 3]
```diff
@@ -1,317 +1,368 @@
...
+
+duk_ret_t duk_get_from_php(duk_context * ctx)
+{
+    char * name = duk_get_string(ctx, 1);
+
+    if (name[0] == '$') {
+        zval ** value;
+        TSRMLS_FETCH();
+        if(zend_hash_find(EG(active_symbol_table), name+1, strlen(name), (void **) &value) == SUCCESS) {
+            zval_to_duk(ctx, NULL, *value);
```
Comment (crodas) [inline phpjs.c:354]: TSRMLS_CC at the end

File: interface.c [hunk 4]
```diff
@@ -1,300 +1,452 @@
...
+    FETCH_THIS_EX(0);
+    obj->ctx = duk_create_heap(NULL, NULL, NULL, getThis(), NULL);
+    duk_php_init(obj);
+    // set API
+    {
+        zval* a_value;
+        zval* require_fn;
+        zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "|zz", &a_value, &require_fn);
+        if(IS_ARRAY ==  Z_TYPE_P(a_value)){
+            zval_to_duk(obj->ctx, "API", a_value);
```
Comment (crodas) [inline interface.c:51]: TSRMLS_CC at the end

File: phpjs.c [hunk 5]
```diff
@@ -1,317 +1,368 @@
...
+        duk_push_null(ctx);
+        break;
+    case IS_RESOURCE:
+        duk_push_pointer(ctx, Z_RESVAL_P(value));
+        break;
+    case IS_OBJECT: {
+            zval *val;
+            MAKE_STD_ZVAL(val);
+            ZVAL_ZVAL(val,value,1,0);
+            if(zend_is_callable(val, 0, NULL TSRMLS_CC)){
```
Comment (crodas) [inline phpjs.c:122]: or you can just ignore all my comments and use `TSRMLS_FETCH();` instead.

**Comment by crodas** [comment]
Awesome job! :+1:  Just add `TSRMLS_FETCH();` in line 121 in `phpjs.c` and it should pass the test.

**Comment by badlee** [comment]
Done. [Here](https://github.com/crodas/PHPJS/pull/2)

**Pull request closed without merging**